### DISCUSSION TITLE
记录那些遇到过的屎山代码问题

### DISCUSSION THREAD

**Discussion by sikouhjw** [sikouhjw/gdutthesis#93] [Show and tell]
所谓『屎山』，就是指在垃圾代码的基础上写垃圾代码，『够用就行』的思想让写代码的人不考虑其它后果，只管生产垃圾代码。

**劝告：不要在所谓『师兄师姐』论文代码的基础上去修改出你的论文代码，因为你的『师兄师姐』大概率写的就是屎山。**

**Comment by OsbertWang** [comment]
然而实际上，够用就行已经成为了许多人心中的典范~

**Comment by sikouhjw** [comment]
# 引用公式的编号错误
![图片](https://user-images.githubusercontent.com/49770925/231747385-227dd1c9-e5d8-4a63-b304-b0a952473c0d.png)

问题在于，所有公式的标签 `\label` 都用相同的，即全部都是 `\label{aaa}` 和 `\label{bbb}`，导致引用 `\ref{aaa}` 输出『最后一处 `\label{aaa}`』的公式编号。

**Comment by sikouhjw** [comment]
# 参考文献引用无法压缩

代码写为 `\cite{a}\cite{b}\cite{c}\cite{d}`，输出 `[58][59][60][61]`。想压缩为 `[58-61]`，实际上输出为 `[58-59]`。

不知道怎么写的，只知道正常写**不可能**出现这种问题。

**Comment by sikouhjw** [comment]
# 新建定理类环境，而不会设置『依赖于』章节编号

代码为 `\newtheorem{   remark}{  注    }`（空格为如实记录，因为模板定义了若干定理类环境，新建 `remark` 会报错），输出为『注 1』，问怎么变成『注 1-1』。

随手搜了一个知乎链接[^1]发过去，很明显原文已经把答案给出来了：
> `\newtheorem{name}[counter]{text}[section]`
> section：定理编号依赖于某个章节层次（比如：定理1.1）

然而他的做法是绕远路，就是不用 `amsthm` 自带的接口：
```tex
\usepackage{amsthm}
\usepackage{chngcntr}
% 定义一个新的计数器
\newcounter{remarkcounter}[chapter]
\counterwithin*{remarkcounter}{chapter}
% 自定义 remark 环境
\newenvironment{remark}
  {\refstepcounter{remarkcounter}\par\noindent\textbf{注\thechapter-\arabic{remarkcounter}：}}
  {\par}  
```
问题在于，一、手写定理类环境，没有统一性，二、不用 `amsthm` 自带的修改计数器依赖的接口。

[^1]: https://zhuanlan.zhihu.com/p/133244838

**Comment by sikouhjw** [comment]
# 乱加载宏包

不懂『宏包功能』就加载，例如 `\usepackage{indentfirst}`，这个宏包是用于段首缩进，而模板原生依赖 `ctexbook` 文档类，天然支持段首缩进，完全不需要额外的缩进宏包。

可见网上的屎山代码害人不浅，一代又一代的传下去，只会让真正的、正确的代码被淹没。

给 ai 喂的是屎，难道还能给你生产出牛奶不成？现在的 ai 关于中文 LaTeX 方面，就是一坨屎。

**Comment by sikouhjw** [comment]
# 没看完 `main.tex` 文件和用户手册就来提问

模板选项 `type` 是用来选择学位类型（博士、学硕、专硕）的：
https://github.com/sikouhjw/gdutthesis/blob/f7e5fd504bb3c220d9708d9573630236ed0d793c/main.tex#L12-L16
具体的学位类型是通过 `info/degree` 来填写的：
https://github.com/sikouhjw/gdutthesis/blob/f7e5fd504bb3c220d9708d9573630236ed0d793c/main.tex#L104-L106

而有人在选择 `type=doctor` 后，来问：为什么选择了 `type=doctor`，『工学硕士』没有变成『工学博士』呢？

**Reply by sikouhjw** [reply]
补充，还有问：该怎么显示专硕封面背景而不是学硕封面背景的。在 `main.tex` 一开头选择 `type=promaster` 就完了的事。

**Comment by sikouhjw** [comment]
# 拿另一个模板来我这问的

发现模板编译不通过，过来线下咨询，随后发现该模板不是本模板。

**Comment by sikouhjw** [comment]
# 不懂最小工作示例，不懂将内容脱敏，万物皆可打码，甚至连参考文献都遮遮掩掩

- 提问不知道给代码，只知道给模糊的、打码后的照片。
- 索要代码就说敏感内容（无非就是投稿、写毕业论文，说句不好听的，有几个人的论文是有含金量的？）。
- 让他给 MWE（最小工作示例）就装离线（因为不想看这么长的文字解释什么是 MWE、没有耐心，我甚至怀疑连阅读理解长文本的能力都没有，怎么读的研？）。
- 让他简单的将内容脱敏（把内容删掉，换成无意义的文字），也没几个会继续回复。
- 内容打码我还能理解，参考文献还打码，遮遮掩掩不给源代码以供复现问题，这是几个意思？怕被人从参考文献找到你做的内容？如果这么短的时间就能找到并做出来，那做的内容是不是太过简单了？

**Comment by sikouhjw** [comment]
# 修改表格的编号格式后问表格编号怎么不连续了

原表格编号定义为
```tex
\numberwithin{table}{section}
\renewcommand\thetable{\arabic{chapter}.\arabic{section}.\arabic{table}}
```
他将编号改为
```tex
\renewcommand\thetable{\arabic{chapter}.\arabic{table}}
```
然后问表格编号怎么不连续了，即『表格1.1』『表格1.2』『表格1.3』（新 section）『表格1.1』。

既自己瞎改，又不说自己改了什么。

**Comment by sikouhjw** [comment]
# 只更新 `main.tex` 而不更新 cls 文件

只更新 `main.tex` 而不同步更新 cls 文件，导致报错。

**Comment by sikouhjw** [comment]
# 听不懂人话

问题为 `ccmap` 宏包报错
![图片](https://user-images.githubusercontent.com/49770925/232664352-a7da1143-b7dc-4d4e-abe1-41cf6e4e87de.png)
A: ccmap是CTeX套装的东西，注释掉
Q: 请问一下  怎么注释掉？
A: %
Q: 好的谢谢 但是好像每一行都这么报错
![图片](https://user-images.githubusercontent.com/49770925/232664988-e050d8aa-6197-4f3c-89a0-52fe89a26189.png)
A: 为什么叫你删掉ccmap你没删
Q: 我不知道怎么删 把这行代码删除吗？

结果删掉的是 texstudio 的『红色提示行』，而没有删掉 ccmap 那一行。

**Comment by sikouhjw** [comment]
# 数数不太好

问题：我数了一下，tabular的列数没错啊，但是我要是再加一列（`|c|`），他就停止报错，但是表格右侧就会多一列空白列。
```tex
\begin{tabular}{|c|c|c|c|c|}
  \hline\hline
  a & b & c & d & e\\
  \hline
  1 & & & & & \\
  \hline
\end{tabular}
```
问题在于第一行只写了五列，第二行却写了六列。

**Comment by sikouhjw** [comment]
# 知网导出的 bib 不规范

`author` 带 `&` 符号，导致编译时报错。
```tex
@article{HSYJ202003011,
 author = {严婷 &amp; 韩汉君},
 title = {中国对外直接投资对母国行业间收入差距的影响},
 journal = {上海经济研究},
 volume = {No.378},
 number = {94-108},
 year = {2020},
 issn = {1005-1309},
 doi ={10.19626/j.cnki.cn31-1163/f.2020.03.007}
 }
```

**Comment by sikouhjw** [comment]
# 沟通不畅且胡乱设置

某群有人用 `latex.exe` 编译，询问为什么没有 pdf 出现，让其用 `xelatex`，其回答：『换了还是一样』。

在群里沟通让他下载向日葵，从而远程协助。在加上好友后，询问向日葵的识别码，回『YX0***』（识别码为纯数字）；询问是否下载了向日葵，回『没有』。

打开 texstudio 的配置后，发现 `xelatex`/`pdflatex`/`latex` 的配置栏均为 `latex.exe`，将全部配置恢复默认后问题解决。

**Comment by sikouhjw** [comment]
# 将 word 思维用到了 latex 并且屡教不改
远程操控示范如何保存 tex 文件以及用 texstudio 编译出 pdf 后，他将 tex 另存为 pdf，然后又来问『无法打开 pdf』、『一直存在这个问题，有解决方案吗』、『我就是把那个tex另存为PDF，然后这个PDF打不开』、『那应该如何导出这个模板的PDF』。
